src/core.js

```diff
@@ -221,9 +224,9 @@ const alignWithDOM = function(nodeName, key) {
 
   // Check to see if the node has moved within the parent.
   if (key) {
-    node = keyMap[key];
-    if (node && process.env.NODE_ENV !== 'production') {
-      assertKeyedTagMatches(getData(node).nodeName, nodeName, key);
+    const keyNode = keyMap[key];
+    if (matches(nodeName, key, keyNode)) {
+      node = keyNode;
     }
```

Comment:
Yah, leaving it in the DOM tree can lead to all sorts of manipulations to get everything in the right spot. We could add an extra conditional to the manipulation logic below to account for a replacement.